921

Chadys opened the pull request:
Axelrod-Python/Axelrod#921
I took more time than I though to correct the tests, since `TestHardGoByMajority` uses `TestGoByMajority`. Tell me if there is a problem with how I changed things.
And I also corrected the type hint of memory_depth which indicated `float` even though, as the docstring says on line 37 :
```
Parameters
        ----------
        memory_depth, int >= 0
```
(default value `float('inf')` must have confused the contributor)

meatballs commented on axelrod/tests/strategies/test_gobymajority.py:
```diff
@@ -74,10 +74,15 @@ def test_strategy(self):
-        opponent = axelrod.Cooperator()
-        self.assertEqual('C', player.strategy(opponent))
+    def test_soft(self):
+        pass
```
Why is this here? Does it need a test to be added?

Chadys replied:
No, it's just to override the test from TestGoByMajority, that checks the soft argument works correctly. Since HardGoByMajority doesn't have a soft argument anymore, that test isn't needed and was causing an exception if I didn't override it to remove it. I thought about not making TestHardGoByMajority inherite from TestGoByMajority at all, but there are still enough common points between the two for it to be meaningful.

meatballs replied:
👍  Could you add a comment in to explain all that?

Chadys replied:
I could have used delattr instead.

meatballs replied:
👍  Could you add a comment to explain the need for line 78? (I had to ask this again to keep github's review process happy).

Chadys replied:
Sure, will do tonight !

Chadys commented:
So, maybe I got a little over-enthusiastic here, but I modified more things than planned. Instead of adding a comment, I finally concluded that, since `TestHardGoByMajority` proceeded to less tests than `TestGoByMajority`, it would be more logical to inverse the genealogy. So that's what I did, making `TestHardGoByMajority` much clearer by not having any function dealing with the `soft` parameter and making only `TestGoByMajority` implements them. I also re-factorized `test_strategy` a little, to avoid code repetition. And finally I added a test that was missing (testing that `GoByMajority` cooperates when opponent cooperated more often than defected, as only the 'equality' or 'more defect' cases were tested). But if I overstepped my bounds please tell me, and I'll just add the comment instead.

meatballs commented:
> But if I overstepped my bounds please tell me

I don't think we have any to be overstepped!

I'll have a proper look at this in the morning.

marcharper commented:
Looks good to me!

drvinceknight commented:
:+1:
